docs/CONTRIBUTING/Dependencies.md

```diff
@@ -98,7 +98,7 @@ You will also need `autotools`. That is usually available by default but might n
 
 To build for Android, you must download components of the Android SDK, the Android NDK, and Java Development Kit.
 
-RenderDoc is currently known to build with NDK 14b, SDK tools 3859397, SDK build-tools 26.0.1, SDK platform android-23, Java 8 (also known as 1.8). If you use different versions of any of these it is up to you to ensure you have compatible versions of all as otherwise build failures may occur as versions of some components may not be compatible with different (even newer) versions of other components.
+RenderDoc is currently known to build with NDK 21d, SDK tools 3859397, SDK build-tools 26.0.1, SDK platform android-30, Java 8 (also known as 1.8). If you use different versions of any of these it is up to you to ensure you have compatible versions of all as otherwise build failures may occur as versions of some components may not be compatible with different (even newer) versions of other components.
```

Comment:
Can we get this to work without changing the SDK versions? I don't want to have to bump the required version as these SDK versions are _incredibly_ fragile and might break on some system.

I guess this is needed for the java code changes since nothing else seems like it would have any dependency. Can we manually forward declare (or dynamically access? I'm not sure how java works) any interfaces we might need, and inline any constants?

Reply:
I found that we can keep the NDK version the same! however I think the android platform level needs to be rolled forward to access the new API features for scoped storage in AndroidManifest.XML and Loader.java (the permissions and new Intents specifically)

To my knowledge, I don't know if any way that we can build against those features independently of the platform level or dynamically access those features unfortunately

Reply:
For the java parts I can't test whether it runs correctly, but it looks like replacing the new constants with their documented values and then using java reflection to access `android.os.Environment.isExternalStorageManager()` at least compiles for me:

```java
        if (Build.VERSION.SDK_INT < 30 /*Build.VERSION_CODES.R*/) {
            permission[0] = android.Manifest.permission.WRITE_EXTERNAL_STORAGE;         

            requestPermissions(permission, 1);

            // Request is asynchronous, so prevent connection to server until permissions granted.
            while(checkSelfPermission(permission[0])
                    != android.content.pm.PackageManager.PERMISSION_GRANTED)
            {
                try {
                    Thread.sleep(1000);
                } catch (InterruptedException e) {
                    break;
                }
            }
        } else {
            permission[0] = "android.permission.MANAGE_EXTERNAL_STORAGE"; // android.Manifest.permission.MANAGE_EXTERNAL_STORAGE;

            try {
                java.lang.reflect.Method method = Environment.class.getMethod("isExternalStorageManager");

                Boolean result = (Boolean)method.invoke(null); 

                // Popup a dialog if we haven't granted Android storage permissions.                
                if (!result) {
                    Intent viewIntent = new Intent( "android.settings.MANAGE_ALL_FILES_ACCESS_PERMISSION"  /*android.provider.Settings.ACTION_MANAGE_ALL_FILES_ACCESS_PERMISSION*/);
                    startActivity(viewIntent);
                }
            } catch(Exception e) {
            }
        }
```

The `android:requestLegacyExternalStorage` application attribute I'm not so sure how to reference on older SDKs. However looking at the documentation for it that says that it's ignored on Android 11 - which seems like it would make sense to me since as I understand it the point of this change is to make RenderDoc's file access work on the new system. Is that attribute still needed, or is it a holdover from earlier experiments?

Reply:
Oh I see! I'm not super familiar with Java Reflections so I didn't know that it worked this way, I tested things on my end and this worked perfectly across the devices I have! (Android 8, Android 10, and Android 11) and that also means we do not have to push the platform version either!

Also you are correct the `android:requestLegacyExternalStorage` was a holdover from when I was trying to use the /media/ path for all versions of Android, so that can go away as well

Thanks for these suggestions!